src/core/positioning/positioningsource.cpp

```diff
@@ -0,0 +1,317 @@
+  }
+  else
+  {
+    mCollectedPositionInformations.clear();
+  }
+
+  emit averagedPositionCountChanged();
+  emit averagedPositionChanged();
+}
+
+void PositioningSource::setLogging( bool logging )
+{
+  if ( mLogging == logging )
+    return;
+
+  mLogging = logging;
+
+  if ( mReceiver )
+  {
+    if ( mLogging )
+    {
+      mReceiver->startLogging();
+    }
+    else
+    {
+      mReceiver->stopLogging();
+    }
+  }
+
+  emit loggingChanged();
+}
+
+void PositioningSource::setElevationCorrectionMode( ElevationCorrectionMode elevationCorrectionMode )
+{
+  if ( mElevationCorrectionMode == elevationCorrectionMode )
+    return;
+
+  mElevationCorrectionMode = elevationCorrectionMode;
+
+  emit elevationCorrectionModeChanged();
+}
+
+void PositioningSource::setAntennaHeight( double antennaHeight )
+{
+  if ( mAntennaHeight == antennaHeight )
+    return;
+
+  mAntennaHeight = antennaHeight;
+
+  emit antennaHeightChanged();
+}
+
+void PositioningSource::setupDevice()
+{
+  if ( mReceiver )
+  {
+    mReceiver->disconnectDevice();
+    mReceiver->stopLogging();
+    disconnect( mReceiver, &AbstractGnssReceiver::lastGnssPositionInformationChanged, this, &PositioningSource::lastGnssPositionInformationChanged );
+    disconnect( mReceiver, &AbstractGnssReceiver::lastErrorChanged, this, &PositioningSource::deviceLastErrorChanged );
+    disconnect( mReceiver, &AbstractGnssReceiver::socketStateChanged, this, &PositioningSource::deviceSocketStateChanged );
+    disconnect( mReceiver, &AbstractGnssReceiver::socketStateStringChanged, this, &PositioningSource::deviceSocketStateStringChanged );
+    mReceiver->deleteLater();
+    mReceiver = nullptr;
+  }
+
+  if ( mDeviceId.isEmpty() )
+  {
+    mReceiver = new InternalGnssReceiver( this );
+  }
+  else
+  {
+    if ( mDeviceId.startsWith( TcpReceiver::identifier + ":" ) )
+    {
+      const qsizetype portSeparator = mDeviceId.lastIndexOf( ':' );
+      const QString address = mDeviceId.mid( 4, portSeparator - 4 );
+      const int port = mDeviceId.mid( portSeparator + 1 ).toInt();
+      mReceiver = new TcpReceiver( address, port, this );
+    }
+    else if ( mDeviceId.startsWith( UdpReceiver::identifier + ":" ) )
+    {
+      const qsizetype portSeparator = mDeviceId.lastIndexOf( ':' );
+      const QString address = mDeviceId.mid( 4, portSeparator - 4 );
+      const int port = mDeviceId.mid( portSeparator + 1 ).toInt();
+      mReceiver = new UdpReceiver( address, port, this );
+    }
+    else if ( mDeviceId.startsWith( EgenioussReceiver::identifier + ":" ) )
+    {
+      mReceiver = new EgenioussReceiver( this );
+    }
+#ifdef WITH_SERIALPORT
+    else if ( mDeviceId.startsWith( SerialPortReceiver::identifier + ":" ) )
+    {
+      const QString address = mDeviceId.mid( 7 );
+      mReceiver = new SerialPortReceiver( address, this );
+    }
+#endif
+    else
+    {
+#ifdef WITH_BLUETOOTH
+      mReceiver = new BluetoothReceiver( mDeviceId, this );
+#endif
+    }
```

Comment:
I guess there is no `else if` condition that can be written, but since we have `ifdef`, would it make sense to log something if the device does not support bluetooth? And also that it is defaulting to bluetooth?

Reply:
Yeah I was thinking about this actually. I'll fix in a follow up PR.